File: Algo/Easy/0169_Majority_Element/Solution.py
from typing import *

class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        cur_majority = nums[0]
        count = 0
        
        for num in nums:
            if cur_majority == num:
                count += 1
            else:
                if count == 0:
                    cur_majority = num
                else:
                    count -= 1
        
        return cur_majority

File: Algo/Easy/0169_Majority_Element/Solution.java
import java.util.*;

// 2021-06-09 (O(n) space)
class Solution {
    public int majorityElement(int[] nums) {
        Map<Integer, Integer> counts = new HashMap<Integer, Integer>();
        counts.put(nums[0], 1);
        int currentMax = nums[0];
        int maxCount = 1;
        
        for (int i = 1; i < nums.length; i++) {
            
            int current = counts.getOrDefault(nums[i], 0);
            counts.put(nums[i], ++current);
            
            if (counts.get(nums[i]) > maxCount) {
                currentMax = nums[i];
                maxCount = counts.get(nums[i]);
            }
                
            if (maxCount > nums.length / 2) break;
        }
        
        return currentMax;
    }
}